### ISSUE TITLE
Packaging fails on linux (with segfault)

### ISSUE THREAD

**Issue by galeone** [adamrehn/ue4-docker#67]
Output of the `ue4-docker info` command:

```
ue4-docker version:         0.0.45 (latest available version is 0.0.50)
Operating system:           Linux (Ubuntu 18.04.4 LTS, 5.3.0-28-generic)
Docker daemon version:      19.03.5
NVIDIA Docker supported:    No
Maximum image size:         No limit detected
Available disk space:       221.06 GiB
Total system memory:        31.26 GiB physical, 2 GiB virtual
Number of processors:       6 physical, 12 logical
```

I'm trying to package a project using ue4docker on Linux. The compilation process works fine, but as soon as the cooking process starts, the process terminates with a segfault. Here's the log:

```
ProcessResult.StdOut:   LogShaderCompilers: Display: Worker (3/13): shaders left to compile 1
ProcessResult.StdOut:   LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_M.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
ProcessResult.StdOut:   LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_M.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
ProcessResult.StdOut:   LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_T.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
ProcessResult.StdOut:   LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_T.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
ProcessResult.StdOut:   LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_TM.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
ProcessResult.StdErr:   [0210/162856:ERROR:browser_main_loop.cc(217)] Running without the SUID sandbox! See https://chromium.googlesource.com/chromium/src/+/master/docs/linux_suid_sandbox_development.md for more information on developing with the sandbox on.
CommandUtils.Run: Took 22.229125s to run UE4Editor, ExitCode=139
InternalUtils.SafeDeleteFile: SafeDeleteFile /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/Cook-2020.02.10-16.28.57.txt
InternalUtils.SafeCopyFile: SafeCopyFile /home/ue4/UnrealEngine/Engine/Programs/AutomationTool/Saved/Cook-2020.02.10-16.28.35.txt /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/Cook-2020.02.10-16.28.57.txt
InternalUtils.SafeDeleteFile: SafeDeleteFile /home/ue4/UnrealEngine/Engine/Programs/AutomationTool/Saved/Cook-2020.02.10-16.28.35.txt
Log.WriteException: ==============================================================================
Log.WriteException: ERROR: Cook failed.
Log.WriteException:        (see /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/Log.txt for full exception trace)
Log.WriteException: 
Log.WriteException: CommandletException: Editor terminated with exit code 139 (segmentation fault) while running Cook for /builds/project-0/Dreamcatcher/Dreamcatcher.uproject; see log /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/Cook-2020.02.10-16.28.57.txt
Log.WriteException:   at AutomationTool.CommandUtils.RunCommandlet (Tools.DotNETCommon.FileReference ProjectName, System.String UE4Exe, System.String Commandlet, System.String Parameters, System.String& DestLogFile) [0x0076d] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.CommandUtils.RunCommandlet (Tools.DotNETCommon.FileReference ProjectName, System.String UE4Exe, System.String Commandlet, System.String Parameters) [0x00001] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.CommandUtils.CookCommandlet (Tools.DotNETCommon.FileReference ProjectName, System.String UE4Exe, System.String[] Maps, System.String[] Dirs, System.String InternationalizationPreset, System.String[] CulturesToCook, System.String TargetPlatform, System.String Parameters) [0x0013e] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at Project.Cook (AutomationTool.ProjectParams Params) [0x007ba] in <ceffa9e3a3664ec9972c2287d36ff5e2>:0 Wrapped by AutomationException: Cook failed.
Log.WriteException:   at Project.Cook (AutomationTool.ProjectParams Params) [0x007f7] in <ceffa9e3a3664ec9972c2287d36ff5e2>:0 
Log.WriteException:   at BuildCookRun.DoBuildCookRun (AutomationTool.ProjectParams Params) [0x00074] in <ceffa9e3a3664ec9972c2287d36ff5e2>:0 
Log.WriteException:   at BuildCookRun.ExecuteBuild () [0x00040] in <ceffa9e3a3664ec9972c2287d36ff5e2>:0 
Log.WriteException:   at AutomationTool.BuildCommand.Execute () [0x00001] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.Automation.Execute (System.Collections.Generic.List`1[T] CommandsToExecute, System.Collections.Generic.Dictionary`2[TKey,TValue] Commands) [0x0007c] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.Automation.Process (System.String[] Arguments, UnrealBuildTool.StartupTraceListener StartupListener) [0x00282] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.Program.MainProc (System.String[] Arguments, UnrealBuildTool.StartupTraceListener StartupListener) [0x00001] in <e50303f0210a434eab4ab3c334bec706>:0 
Log.WriteException:   at AutomationTool.Program+<>c__DisplayClass1_0.<Main>b__2 () [0x00000] in <e50303f0210a434eab4ab3c334bec706>:0 
Log.WriteException:   at AutomationTool.InternalUtils.RunSingleInstance (System.Func`1[TResult] Main) [0x000a0] in <24a147dd91b247e1ac4c3b43a96c96e3>:0 
Log.WriteException:   at AutomationTool.Program.Main (System.String[] Arguments) [0x0020a] in <e50303f0210a434eab4ab3c334bec706>:0
Log.WriteException: ==============================================================================
Program.Main: AutomationTool exiting with ExitCode=25 (Error_UnknownCookFailure)

```

Looking at the more detailed log `Cook-2020.02.10-16.28.35.txt ` these are the last lines (not very useful):

```
LogMaterial: Display: Missing cached shader map for material DefaultSpriteMaterial, compiling. 
LogMaterial: Display: Missing cached shader map for material DefaultSpriteMaterial, compiling. 
```

On windows, I can package this very same project correctly (using the windows containers), so my guess is that there is some missing library in the Linux containes, but I don't know where to look honestly. Moreover, using `ue4 package` from the Linux host (and not from the Linux container) I'm able to package the project correctly.
Any help is appreciated.

Thank you!

**Comment by adamrehn** [comment]
The required dependencies for CEF are now included in the ue4-full image under Linux as of ue4-docker version 0.0.51, could you please test to see if that fixes your segfault?

**Comment by galeone** [comment]
I was using a container with all the CEF dependencies already installed, and it still segfaults.
The strange fact is that it looks like it segfaults when it should start the cooking process and thus, I don't know if it is something related to the cooker itself, or to the unreal project I'm building.

**Comment by adamrehn** [comment]
Hmmm, the mystery deepens. Could you please post the full contents of the log file? I know that cooking often uses multiple threads to run tasks in parallel, so it's possible that there's a clue earlier in the output that might help diagnose the cause of the failure.

**Comment by galeone** [comment]
Sure, here are the full logs [log.zip](https://github.com/adamrehn/ue4-docker/files/4190507/log.zip)

**Comment by adamrehn** [comment]
Nothing is jumping out as suspicious in those log files. It might be worth running the cook command through a debugger. If you install `gdb` in the container image and run a container with the `--priveleged` flag, you should be able to debug it and produce a backtrace like so:

```
$ gdb /home/ue4/UnrealEngine/Engine/Binaries/Linux/UE4Editor

(gdb) run /builds/project-0/Dreamcatcher/Dreamcatcher.uproject -run=Cook -TargetPlatform=LinuxNoEditor -fileopenlog -unversioned -abslog=/home/ue4/UnrealEngine/Engine/Programs/AutomationTool/Saved/Cook-2020.02.10-16.28.35.txt -stdout -CrashForUAT -unattended -NoLogTimes -UTF8Output
(gdb) bt
```

**Comment by galeone** [comment]
It is still `libcef.so`!

```
0x00007fff7e760533 in base::ThreadTaskRunnerHandle::Get() ()
   from /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
(gdb) bt
#0  0x00007fff7e760533 in base::ThreadTaskRunnerHandle::Get() ()
    at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#1  0x00007fff7e762efa in base::SequencedWorkerPool::SequencedWorkerPool(unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) ()
    at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#2  0x00007fff81142dab in base::LazyInstance<content::(anonymous namespace)::BrowserThreadGlobals, base::internal::LeakyLazyInstanceTraits<content::(anonymous namespace)::BrowserThreadGlobals> >::Get() ()
    at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#3  0x00007fff8114370a in content::BrowserThread::CurrentlyOn(content::BrowserThread::ID) ()
    at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#4  0x00007fff7e64ac46 in CefContext::Initialize(CefMainArgs const&, CefStructBase<CefSettingsTraits> const&, CefRefPtr<CefApp>, void*) () at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#5  0x00007fff7e64aa1b in CefInitialize(CefMainArgs const&, CefStructBase<CefSettingsTraits> const&, CefRefPtr<CefApp>, void*) ()
    at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#6  0x00007fff7e5fce78 in cef_initialize () at /home/ue4/UnrealEngine/Engine/Binaries/Linux/../ThirdParty/CEF3/Linux/libcef.so
#7  0x00007fff82f55c7b in CefInitialize(CefMainArgs const&, CefStructBase<CefSettingsTraits> const&, CefRefPtr<CefApp>, void*) ()
    at Runtime/Core/Public/Delegates/DelegateSignatureImpl.inl:343
#8  0x00007fff82f131fd in FWebBrowserSingleton::FWebBrowserSingleton(FWebBrowserInitSettings const&) (this=<optimized out>, WebBrowserInitSettings=...) at Runtime/WebBrowser/Private/WebBrowserSingleton.cpp:316
#9  0x00007fff82f11759 in FWebBrowserModule::GetSingleton() (this=<optimized out>)
    at Runtime/WebBrowser/Private/WebBrowserModule.cpp:69
#10 0x00007fff83402cfa in FWebBrowserWidgetModule::StartupModule() (this=0x7fff9a7cdc00)
    at /home/ue4/UnrealEngine/Engine/Plugins/Runtime/WebBrowserWidget/Source/WebBrowserWidget/Private/WebBrowserWidgetModule.cpp:23
#11 0x00007ffff7193dc5 in FModuleManager::LoadModuleWithFailureReason(FName, EModuleLoadResult&) (this=<optimized out>, InModuleName=..., OutFailureReason=<optimized out>) at Runtime/Core/Private/Modules/ModuleManager.cpp:548
#12 0x00007ffff1dac692 in FModuleDescriptor::LoadModulesForPhase(ELoadingPhase::Type, TArray<FModuleDescriptor, TSizedDefaultAllocator<32> > const&, TMap<FName, EModuleLoadResult, FDefaultSetAllocator, TDefaultMapHashableKeyFuncs<FName, EModuleLoadResult, false> >&) (LoadingPhase=ELoadingPhase::PreDefault, Modules=..., ModuleLoadErrors=...) at Runtime/Projects/Private/ModuleDescriptor.cpp:595
#13 0x00007ffff1dbc4a3 in TryLoadModulesForPlugin(FPlugin const&, ELoadingPhase::Type) (Plugin=..., LoadingPhase=38)
    at Runtime/Projects/Private/PluginManager.cpp:936
#14 0x00007ffff1dbbf47 in FPluginManager::LoadModulesForEnabledPlugins(ELoadingPhase::Type) (this=<optimized out>, LoadingPhase=ELoadingPhase::PreDefault) at Runtime/Projects/Private/PluginManager.cpp:1003
#15 0x0000000000251003 in FEngineLoop::LoadStartupModules() (this=<optimized out>)
    at Runtime/Launch/Private/LaunchEngineLoop.cpp:3337
#16 0x000000000022fb3c in FEngineLoop::PreInit(char16_t const*) (this=<optimized out>, CmdLine=<optimized out>)
    at Runtime/Launch/Private/LaunchEngineLoop.cpp:2738
#17 0x0000000000248f5c in EnginePreInit(char16_t const*) (CmdLine=<optimized out>) at Runtime/Launch/Private/Launch.cpp:42
#18 0x0000000000248f5c in GuardedMain(char16_t const*) (CmdLine=0x7fffcfb63a80 u" /builds/project-0/Dreamcatcher/Dreamcatcher.uproject -run=Cook -TargetPlatform=LinuxNoEditor -fileopenlog -unversioned -abslog=/home/ue4/UnrealEngine/Engine/Programs/AutomationTool/Saved/Cook-2020.02"...) at Runtime/Launch/Private/Launch.cpp:131
#19 0x00007fffece04772 in CommonUnixMain(int, char**, int (*)(char16_t const*), void (*)()) (argc=<optimized out>, argv=0x7fffffffe2d8, RealMain=0x248f00 <GuardedMain(char16_t const*)>, AppExitCallback=0x254170 <LaunchUnix_FEngineLoop_AppExit()>)
    at Runtime/Unix/UnixCommonStartup/Private/UnixCommonStartup.cpp:239
#20 0x00007fffec82ab97 in __libc_start_main () at /lib/x86_64-linux-gnu/libc.so.6
#21 0x000000000022a029 in _start ()
```

I've already solved this issue in this past (I'm going to test my solution and share it with you the result): in short, cef **requires** a graphical interface *running*. Thus, I work around this by executing `xvfb` before running the command.
I'll try to install it into the container and do this - if it works, perhaps it is worth it to ship `xvfb` inside the container

**Comment by galeone** [comment]
Yes, the segfault disappears by running `Xvfb :94 -screen 0 1280x1024x24 -ac -nolisten tcp -nolisten unix` before the cooking step.

Now, at least I don't have a segfault, but I still cannot execute the cooking process (perhaps because of missing runtime libraries I have to install in the system, but that are not related to unreal, but related to my specific project)

**Comment by adamrehn** [comment]
That's interesting, I already knew that CEF requires an X server to be running in order to display a WebBrowserWidget when a map is loaded, but I didn't think it was necessary for simply cooking a project that uses CEF.

**Comment by adamrehn** [comment]
Okay, I've managed to reproduce the error with a minimal test project, and the source code for the WebBrowserWidget module [definitely initialises CEF on startup](https://github.com/EpicGames/UnrealEngine/blob/4.24.2-release/Engine/Plugins/Runtime/WebBrowserWidget/Source/WebBrowserWidget/Private/WebBrowserWidgetModule.cpp#L23), so I suppose it is indeed necessary to include `xvfb` for scenarios where an X11 socket is not being bind-mounted from the host system. I've added it to the ue4-full image in commit af521fe, which is live in ue4-docker version 0.0.52.

I'll have a think about ways that it might be possible to automate creating the virtual display as needed, to avoid users having to perform that step manually.

**Comment by Lance-Gao** [comment]
hi,adamereh, when running the binary with WebBrowserWidget in docker, do I need to run xvfb firstly?@adamrehn

**Comment by galeone** [comment]
Hi @Lance-Gao , you should overwrite the `ENTRYPOINT` of the container, starting `xvfb` and sending it to background (e.g. `xvfb &`).

**Comment by Lance-Gao** [comment]
@galeone I run xvfb firstly(my command is "Xvfb :94 -screen 0 1280x1024x24 -ac -nolisten tcp -nolisten unix > /dev/null 2>&1 &; export DISPLAY=:94"), and then run binary with WebBrowserWidget in docker.
Program can not run successfully. 
Any help is appreciated.Thanks.

**Comment by Lance-Gao** [comment]
Program print 
"4.23.1-8386587+++UE4+Release-4.23 517 0
Disabling core dumps.",
then no other output.

**Comment by Lance-Gao** [comment]
I can not resolve the problem, then I give up the webbroserwidget.

**Comment by adamrehn** [comment]
I'm going to assign this issue to myself as a reminder that I need to add some sort of support for automatically running Xvfb to [ue4cli](https://github.com/adamrehn/ue4cli) and probably also create a demo Dockerfile in the [ue4-example-dockerfiles](https://github.com/adamrehn/ue4-example-dockerfiles) repository to act as an example for developers who want to run Xvfb themselves without using ue4cli.

**Comment by galeone** [comment]
> I'm going to assign this issue to myself as a reminder that I need to add some sort of support for automatically running Xvfb to [ue4cli](https://github.com/adamrehn/ue4cli) and probably also create a demo Dockerfile in the [ue4-example-dockerfiles](https://github.com/adamrehn/ue4-example-dockerfiles) repository to act as an example for developers who want to run Xvfb themselves without using ue4cli.

Why automatically running Xvfb with ue4cli? Isn't the ue4cli just wrapping the ue4 editor and it's independent of the running environment (with/without graphical framebuffer)?
For the docker part, I did something similar in the [tests](https://pgaleone.eu/cicd/unreal-engine/2020/09/30/continuous-integration-with-unreal-engine-4/#tests) section of this article - you can reuse that code in your ue4-example-dockerfiles if you want :+1:

**Comment by adamrehn** [comment]
> Why automatically running Xvfb with ue4cli? Isn't the ue4cli just wrapping the ue4 editor and it's independent of the running environment (with/without graphical framebuffer)?

You're right, it's not really part of ue4cli's job to manage the environment, but I figure that automatically running Xvfb when it detects that you're packaging a project that uses the WebBrowserWidget without a `DISPLAY` would make life easier for developers who encounter this issue. I don't think I'd want to enable the behaviour by default though, probably just emit a warning providing the user with information about the problem and its solution unless they have specified a flag that enables the automatic behaviour.

> For the docker part, I did something similar in the tests section of this article - you can reuse that code in your ue4-example-dockerfiles if you want

Awesome, thanks! I'll be sure to credit you and link back to your blog post when I do. :smile:

**Comment by vivienneanthony** [comment]
Hello. I came upon this thread and I am wondering if this is related. I upgraded my OS from Ubuntu 16.04 to 20.04 onto a new ssd but after the upgrade. I could not cook a project. Hard drives, memory, and video card is fine. After two weeks, I switch back my old ssd with 16.04. The projects built fine. There obvious problem other then it's a issue between Ubuntu 20.04 and Unreal Engine. I am getting the same errors to the first note.  Here are the full logs of the unsuccessful build and cook. Maybe someone can spot a issue. I don't use WebBrowserWidget but maybe theres no display and WebBrowserWidget  is set causing issues??

https://drive.google.com/drive/folders/1TdgEP0mK0fUOBd9DLvCb9OlbDjxVxAMJ?usp=sharing

**Comment by adamrehn** [comment]
@vivienneanthony ah yes, I remember seeing the discussion of your bizarre cook failures a few days ago in the `#linux` channel of the Unreal Slackers Discord server. I can't see any indication that WebBrowserWidget is involved in your crashes, so it's certainly not the problem reported in this thread. Given that you've been attempting to determine whether the OS version (16.04 .vs. 20.04) is the underlying factor or whether some other element is at play, it might be worthwhile using ue4-docker to build a container image for Unreal Engine 4.21 and incorporating the container into your testing. The images produced by ue4-docker are based on Ubuntu 18.04, but also create an Installed Build of the Engine that can be exported to the host system, and you could leverage that fact to perform some controlled tests.

Here's an example workflow for how this might work:

1. **Build container images for Unreal Engine 4.21 using ue4-docker:**
    
    ```bash
    ue4-docker build 4.21.2 --no-engine
    ```
    
    This will produce a number of container images, including one tagged `adamrehn/ue4-full:4.21.2-opengl`.

2. **Attempt to cook your project in a container based on the image you just built:**
    
    ```bash
    # Start a container using the ue4-full image, with the OrvilleGalaxy project directory bind-mounted as `/project`
    docker run --rm -ti -v /path/to/OrvilleGalaxy:/project -w /project "adamrehn/ue4-full:4.21.2-opengl" bash
    
    # Once the container starts, clean any existing build artifacts and attempt to package the project
    ue4 clean
    ue4 package
    ```
    
3. **Export the Unreal Engine installation to the host so that it can be used outside of a container:**
    
    ```bash
    # Export the Installed Build of the Unreal Engine from the container image to a directory on the host user's desktop
    ue4-docker export installed "adamrehn/ue4-full:4.21.2-opengl" ~/Desktop/UnrealExported
    ```
    
4. **Attempt to cook your project on the host system using the exported Unreal Engine installation:**
    
    ```bash
    # Install ue4cli so we can use the `ue4 clean` and `ue4 package` commands
    sudo pip3 install ue4cli
    
    # Point ue4cli to our exported Unreal Engine installation
    ue4 setroot ~/Desktop/UnrealExported
    
    # Change to the OrvilleGalaxy project directory, clean any existing build artifacts and attempt to package the project
    cd /path/to/OrvilleGalaxy
    rm -r -f ./dist
    ue4 clean
    ue4 package
    ```

This way, you'll be able to test the exact same Unreal Engine installation in an Ubuntu 18.04 container and on whichever host version you like (16.04 or 20.04.)

**Comment by avg07** [comment]
Hi, @adamrehn. I am getting the same error when trying to build the project. `ue4-docker version:` `0.0.76`
```
  LogLinker: Warning: Asset '../../Plugins/Runtime/WebBrowserWidget/Content/WebTexture_TM.uasset' has been saved with empty engine version. The asset will be loaded but may be incompatible.
  [0201/072330:ERROR:browser_main_loop.cc(217)] Running without the SUID sandbox! See https://chromium.googlesource.com/chromium/src/+/master/docs/linux_suid_sandbox_development.md for more information on developing with the sandbox on.
Took 127.845909s to run UE4Editor, ExitCode=139
ERROR: Cook failed.
       (see /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/Log.txt for full exception trace)
AutomationTool exiting with ExitCode=25 (Error_UnknownCookFailure)
RunUAT ERROR: AutomationTool was unable to run successfully.
Using user-specified engine root: /home/ue4/UnrealEngine
Traceback (most recent call last):
  File "/usr/local/bin/ue4", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/cli.py", line 222, in main
    SUPPORTED_COMMANDS[command]['action'](manager, args)
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/cli.py", line 94, in <lambda>
    'action': lambda m, args: m.packageDescriptor(os.getcwd(), args),
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/UnrealManagerBase.py", line 452, in packageDescriptor
    self.packageProject(dir, args[0] if len(args) > 0 else 'Shipping', args[1:])
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/UnrealManagerBase.py", line 427, in packageProject
    ] + extraArgs)
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/UnrealManagerBase.py", line 371, in runUAT
    Utility.run([self.getRunUATScript()] + args, cwd=self.getEngineRoot(), raiseOnError=True)
  File "/usr/local/lib/python3.6/dist-packages/ue4cli/Utility.py", line 145, in run
    raise Exception('child process ' + str(command) + ' failed with exit code ' + str(returncode))
Exception: child process ['/home/ue4/UnrealEngine/Engine/Build/BatchFiles/RunUAT.sh', 'BuildCookRun', '-utf8output', '-clientconfig=Development', '-serverconfig=Development', '-project=/tmp/project/TevaZ4.uproject', '-noP4', '-cook', '-build', '-stage', '-prereqs', '-pak', '-archive', '-archivedirectory=/tmp/project/dist', '-platform=Linux', '-allmaps'] failed with exit code 25
The command '/bin/sh -c ue4 package Development' returned a non-zero code: 1
```

I didn't understand how to use Xvfb to ue4-package the project. Could you tell me how to proceed?

**Comment by adamrehn** [comment]
@avg07 please see @galeone's comment for details on the appropriate command to run Xvfb: https://github.com/adamrehn/ue4-docker/issues/67#issuecomment-585124890. You will need to run this command and set the `DISPLAY` environment variable prior to running the [ue4 package](https://docs.adamrehn.com/ue4cli/descriptor-commands/package) command, either by creating a Dockerfile that sets an appropriate `ENTRYPOINT` script or by running the commands interactively inside a container.

**Comment by avg07** [comment]
@adamrehn, thanks for the answer. I have already solved the problem. I have installed xvfb-run in a runtime-сontainer and added it with the run parameter in my k8s pod:
```
command: ["/bin/bash", "-c", "xvfb-run -s '-screen 0 1920x1080x24 -nocursor' 2> /dev/null /home/ue4/project/MyProject.sh -PixelStreamingIP=127.0.0.1 -PixelStreamingPort=8888 -AudioMixer -NvEncFrameRateNum=1 -RenderOffScreen"]
```

Another way is to use xvfb sidecar-container in pod:
```
...
      containers:
        - name: app
          image: "myproject:2"
          env:
            - name: DISPLAY
              value: "localhost:1.0"
          args: ["-PixelStreamingIP=127.0.0.1", "-PixelStreamingPort=8888", "-AudioMixer", "-NvEncFrameRateNum=1", "-RenderOffScreen"]
          volumeMounts:
            - mountPath: /run/user/1000/pulse/native
              name: pulse-native
            - mountPath: /tmp/.X11-unix
              name: x11-unix
              readOnly: false
        - name: xvfb
          image: comiq/xvfb:latest
          env:
            - name: DISPLAY
              value: "1"
            - name: SCREEN
              value: "0"
          ports:
            - name: xserver
              containerPort: 6001
...
```

**Comment by slonopotamus** [comment]
I don't see what we can do here. It's user responsibility to run Xvfb when they build/run project that uses CEF. Someone could try to file a bugreport to CEF (or UE, depending on how they use CEF) so it stops requiring a running X server until user tries to actually render something, but anyway it is out of scope of ue4-docker.

@adamrehn any objections against closing this?

**Comment by adamrehn** [comment]
@slonopotamus I'm just going to investigate one little thing before we call wraps on this one, I'll let you know once I have results to share.

**Comment by adamrehn** [comment]
Right, this is now fixed upstream in the Unreal Engine: https://github.com/EpicGames/UnrealEngine/commit/60c6833e4f92bad80b76cf0e28d96f5ba1dc4c26
